### PR TITLE
[core/docs] Update worker docstring

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?

Add more comprehensive docstring for core remote function.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/worker.py [hunk 1]
```diff
@@ -2777,22 +2778,61 @@ def remote(
...
+    *args, **kwargs
+) -> Union[ray.remote_function.RemoteFunction, ray.actor.ActorClass]:
     """Defines a remote function or an actor class.
 
-    This can be used with no arguments to define a remote function or actor as
-    follows:
+    This function can be used as a decorator with no arguments
+    to define a remote function or actor as follows:
 
     .. code-block:: python
```
Comment (pcmoritz) [inline python/ray/_private/worker.py:2789]: Can we make sure these docstrings run in the CI? See https://github.com/ray-project/ray/pull/28421

Comment (pcmoritz) [reply python/ray/_private/worker.py:2789]: Basically you need to replace `.. code-block:: python` with `>>> ` in front of every line -- examples in the linked PR

File: python/ray/_private/worker.py [hunk 2]
```diff
@@ -2816,18 +2856,49 @@ def method(self):
...
                 return 1
-        Bar = Foo.options(num_cpus=1, resources=None)
+
+        Foo_with_no_resources = Foo.options(num_cpus=1, resources=None)
+        foo_actor = Foo_with_no_resources.remote()
+        assert ray.get(foo_actor.method.remote()) == 1
+
 
     Running remote actors will be terminated when the actor handle to them
     in Python is deleted, which will cause them to complete any outstanding
```
Comment (jjyao) [inline python/ray/_private/worker.py:2876]: When all actor handles to them are deleted

File: python/ray/_private/worker.py [hunk 3]
```diff
@@ -2816,18 +2856,49 @@ def method(self):
...
+    calling ``del actor`` *could* trigger actor deletion. Note that your program
+    may have multiple references to the same ActorHandle, and actor termination
+    will not occur until the reference count goes to 0. See the Python
+    documentation for more context about object deletion.
+    https://docs.python.org/3.9/reference/datamodel.html#object.__del__
+
+    If you want to kill actors immediately, you can also call ``ray.kill(actor)``.
+
+    .. tip::
+        Avoid repeatedly passing in large arguments to remote task or method calls.
```
Comment (jjyao) [inline python/ray/_private/worker.py:2887]: Can we convert it to a Ray anti-pattern and link here?

Or you can add a TODO and I'll do it.

**Review by pcmoritz — APPROVED** [review]
This is now passing the CI -- I still had to skip a small number of doctests because we currently don't have custom resource handling in `make doctest` :)

File: python/ray/_private/worker.py [hunk 4]
```diff
@@ -2587,6 +2588,9 @@ def _mode(worker=global_worker):
 
 
 def _make_remote(function_or_class, options):
+    if not function_or_class.__module__:
+        function_or_class.__module__ = "global"
```
Comment (pcmoritz) [inline python/ray/_private/worker.py:2592]: This is a bit sketch -- it is fixing https://github.com/ray-project/ray/issues/4758 but most likely there is a better fix :)

File: python/ray/_private/worker.py (hunk 3 above)
Comment (richardliaw) [reply python/ray/_private/worker.py:2883]: done

**Pull request merged**